src/Compiler/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorBuilder.cs

```diff
@@ -1,63 +1,313 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
-
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.Extensions.ObjectPool;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-public abstract class BoundAttributeDescriptorBuilder
+public partial class BoundAttributeDescriptorBuilder : TagHelperObject, IBuilder<BoundAttributeDescriptor>
 {
-    public abstract string Name { get; set; }
+    private static readonly ObjectPool<BoundAttributeDescriptorBuilder> s_pool = DefaultPool.Create(Policy.Instance);
 
-    public abstract string TypeName { get; set; }
+    internal static BoundAttributeDescriptorBuilder GetInstance(TagHelperDescriptorBuilder parent, string kind)
+    {
+        var builder = s_pool.Get();
 
-    public abstract bool IsEnum { get; set; }
+        builder._parent = parent;
+        builder._kind = kind;
 
-    public abstract bool IsDictionary { get; set; }
+        return builder;
+    }
 
-    public abstract string IndexerAttributeNamePrefix { get; set; }
+    internal static void ReturnInstance(BoundAttributeDescriptorBuilder builder)
+        => s_pool.Return(builder);
 
-    public abstract string IndexerValueTypeName { get; set; }
+    private static readonly ObjectPool<HashSet<BoundAttributeParameterDescriptor>> s_boundAttributeParameterSetPool
+        = HashSetPool<BoundAttributeParameterDescriptor>.Create(BoundAttributeParameterDescriptorComparer.Default);
 
-    public abstract string Documentation { get; set; }
+    // PERF: A Dictionary<string, string> is used intentionally here for faster lookup over ImmutableDictionary<string, string>.
```

Comment:
Do we have access to ImmutableSegmentedDictionary here? If so, I'd suggest moving to it. Otherwise, we can replace with `FrozenDictionary` when we get SCI 8.

Reply:
Unfortunately, we do not have access to `ImmutableSegmentedDictionary`. It's internal to Roslyn. IMO, this and other collections needs to be properly moved to libraries that can be used elsewhere.

Reply:
Ah, bummer. I can't use `FrozenDictionary` because MS.ANC.Razor.Language isn't mutli-targeted. It's `netstandard2.0` or bust. Sigh... 😞

Reply:
I meant more when VS updates to SCI 8.0 and we can consume that package. That package is multitargeted to netstandard2.0, so we'll get it eventually.

Reply:
Yeah, I just double-checked the netstandard2.0 binary in SCI 8, and `FrozenDictionary` is there and waiting for us. :smile: